Initial FreeBSD support

ppenzin opened the pull request:
Initial FreeBSD support with `EMBED_ICU`, which disables native tests. ICU version in FreeBSD package tree is too new - something to sort out next.

To do

- [ ] PAL synchronization seems to be broken
- [ ] Style errors (copyright headers)

ppenzin committed:
Define BSD regs macros for FreeBSD

ppenzin committed:
FreeBSD: disable PAL_GLOBAL for now (TODO)

ppenzin committed:
Detect FreeBSD in top-level CMake

ppenzin committed:
FreeBSD: locate sysconf

rhuanjl commented:
I'm confused by some of the test fails - does a clean PR fail at the moment (due to a system update?) or has this PR got some unintended side effect?

ShortDevelopment commented:
@rhuanjl It seems like MacOS builds currently fail both on azure and cirrus (Even without changes)
See #7007

ppenzin commented:
That seems to be the case, I haven't looked into why MacOS failed yet, and style failed because I touched PAL.

ppenzin committed:
Include machine/fpu.h on FreeBSD

ppenzin committed:
Add FreeBSD 13+ byteswap.h (TODO)

ppenzin commented:
With these changes (there are a couple of things needing better checks or further investigation) my build hits a wall with system libicu, which is a lot newer than Linux system version or the one we embed.

ppenzin committed:
[FreeBSD] Embed ICU

Use 'gmake' instead of 'make' on FreeBSD, confirm it is available in CMake.

ppenzin committed:
[FreeBSD] Get virtual memory limit

Reuse existing implementation, but change the include from sysctl.h to
resource.h - that should still work on other POSIX platforms.

ppenzin committed:
Access sysinfo on FreeBSD

ppenzin committed:
FreeBSD: pickup GNU compat symbols

Find symbols for GNU gettext and sysinfo. Requires `gettext` and
`libsysinfo` packages (ports) to be installed.

ppenzin committed:
FreeBSD: find execinfo library

ppenzin commented:
Unfortunately, there are still PAL runtime errors on FreeBSD, maybe still too early.

rhuanjl commented:
The idea with CC's PAL integration was to try to only include the parts of PAL we actually need (several components are deliberately omitted) whilst we do have some excess bits I wonder if the issue here is that we are missing code that's needed for FreeBSD

pakstanzndabad62-sys reviewed (changes requested):
All permission controller
